src/Cryptol/Eval/SBV.hs

```diff
@@ -0,0 +1,764 @@
+
+  wordBit _ x idx = pure $! svTestBit x (intSizeOf x - 1 - fromInteger idx)
+
+  wordUpdate _ x idx b = pure $! svSymbolicMerge (kindOf x) False b wtrue wfalse
+    where
+     i' = intSizeOf x - 1 - fromInteger idx
+     wtrue  = x `svOr`  svInteger (kindOf x) (bit i' :: Integer)
+     wfalse = x `svAnd` svInteger (kindOf x) (complement (bit i' :: Integer))
+
+  packWord _ bs  = pure $! packSBV bs
+  unpackWord _ x = pure $! unpackSBV x
+
+  wordEq _ x y = pure $! svEqual x y
+  wordLessThan _ x y = pure $! svLessThan x y
+  wordGreaterThan _ x y = pure $! svGreaterThan x y
+
+  wordSignedLessThan _ x y = pure $! svLessThan sx sy
+    where sx = svSign x
+          sy = svSign y
+
+  joinWord _ x y = pure $! svJoin x y
+
+  splitWord _ _leftW rightW w = pure
+    ( svExtract (intSizeOf w - 1) (fromInteger rightW) w
+    , svExtract (fromInteger rightW - 1) 0 w
+    )
+
+  extractWord _ len start w =
+    pure $! svExtract (fromInteger start + fromInteger len - 1) (fromInteger start) w
+
+  wordAnd _ a b = pure $! svAnd a b
+  wordOr  _ a b = pure $! svOr a b
+  wordXor _ a b = pure $! svXOr a b
+  wordComplement _ a = pure $! svNot a
+
+  wordPlus  _ a b = pure $! svPlus a b
+  wordMinus _ a b = pure $! svMinus a b
+  wordMult  _ a b = pure $! svTimes a b
+  wordNegate _ a  = pure $! svUNeg a
+
+  wordDiv sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svQuot a b  -- TODO! Fix this: see issue #662
+
+  wordMod sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svRem a b   -- TODO! Fix this: see issue #662
+
+  wordSignedDiv sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! signedQuot a b
+
+  wordSignedMod sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! signedRem a b
+
+  wordExp _ a b = sExp a b
+  wordLg2 _ a = sLg2 a
+
+  wordToInt _ x = pure $! svToInteger x
+  wordFromInt _ w i = pure $! svFromInteger w i
+
+  intEq _ a b = pure $! svEqual a b
+  intLessThan _ a b = pure $! svLessThan a b
+  intGreaterThan _ a b = pure $! svGreaterThan a b
+
+  intPlus  _ a b = pure $! svPlus a b
+  intMinus _ a b = pure $! svMinus a b
+  intMult  _ a b = pure $! svTimes a b
+  intNegate _ a  = pure $! SBV.svUNeg a
+  intLg2 _ a = svLg2 a
+
+  intDiv sym a b =
+    do let z = svInteger KUnbounded 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svQuot a b  -- TODO! Fix this: see issue #662
+  intMod sym a b =
+    do let z = svInteger KUnbounded 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svRem a b   -- TODO! Fix this: see issue #662
+
+  intExp sym  a b =
+    do let z = svInteger KUnbounded 0
```

Comment:
Probably should explicitly check for concrete exponent here and use `raiseError` rather than letting SBV throw an exception